src/lib/xmlparse.rs

```diff
@@ -1119,12 +1121,40 @@ pub struct DTD {
     pub paramEntities: HashMap<HashKey, Box<ENTITY>>,
     pub defaultPrefix: PREFIX,
     pub in_eldecl: XML_Bool,
-    pub scaffold: *mut CONTENT_SCAFFOLD,
+    pub scaffold: Cow<'scf, RefCell<Scaffold>>,
     pub contentStringLen: c_uint,
-    pub scaffSize: c_uint,
-    pub scaffCount: c_uint,
-    pub scaffLevel: c_int,
-    pub scaffIndex: *mut c_int,
+}
+
+#[repr(C)]
+#[derive(Clone, Default)]
+pub struct Scaffold {
+    scaffold: Vec<CONTENT_SCAFFOLD>,
+    index: Vec<usize>,
+}
+
+impl Scaffold {
+    fn next_part(&mut self) -> Option<usize> {
```

Comment:
This looks like it could be an iterator

Reply:
Sure, but here `None` can also mean _memory allocation failure_, whereas for `Iterator` it usually means _end of iterator_. Plus this would be more like an infinite mutation iterator, since it appends a new element on each call.